numpy/_core/tests/test_stringdtype.py

```diff
@@ -225,7 +225,22 @@ def test_self_casts(dtype, dtype2, strings):
     else:
         arr.astype(dtype2, casting="safe")
 
-    assert_array_equal(arr[:-1], newarr[:-1])
+    if hasattr(dtype, "na_object") and hasattr(dtype2, "na_object"):
+        na1 = dtype.na_object
+        na2 = dtype2.na_object
+        if na1 is na2:
+            assert_array_equal(arr[:-1], newarr[:-1])
+        else:
+            # comparisons between arrays with distinct NA objects
+            # aren't allowed
+            if ((na1 is pd_NA or na2 is pd_NA or
```

Comment:
How about writing 
```
if ((na1 := getattr(dtype, "na_object", None) is not None
    and (na2 := getattr(dtype2, "na_object", None) is not None
    and (na1 is not na2)
    and (na1 is pd_NA or ...)):
    with pytest.raises
        ...
else:
    assert_array_equal(...)
```

Reply:
That doesn't work because `na_object` being unset is the default. It's only a valid attribute if you pass it:

```
>>> np.dtypes.StringDType().na_object
AttributeError: 'numpy.dtypes.StringDType' object has no attribute 'na_object'.
```

but I see what you're saying, refactored a bit to hopefully be less confusing